sycl/test-e2e/VirtualFunctions/helpers.hpp

```diff
@@ -0,0 +1,53 @@
+#include <algorithm>
+#include <type_traits>
+
+// TODO: strictly speaking, selecting a max alignment here may not be always
+// valid, but for test cases that we have now we expect alignment of all types
+// to be the same.
+// std::aligned_storage uses double under the hood which prevents us from
+// using it on some HW. Therefore we use a custom implementation.
+template <typename... T> struct aligned_storage {
+  static constexpr size_t Len = std::max({sizeof(T)...});
+  static constexpr size_t Align = std::max({alignof(T)...});
+
+  struct type {
+    alignas(Align) unsigned char data[Len];
+  };
+};
+
+// Helper data structure that automatically creates a right (in terms of size
+// and alignment) storage to accomodate a value of any of types T...
+template <typename... T> struct obj_storage_t {
+  static_assert(std::max({alignof(T)...}) == std::min({alignof(T)...}),
+                "Unsupported alignment of input types");
+  using type = typename aligned_storage<T...>::type;
+  static constexpr size_t size = std::max({sizeof(T)...});
+
+  type storage;
+
+  template <typename RetT> RetT *construct(const unsigned int TypeIndex) {
+    if (TypeIndex >= sizeof...(T)) {
+#ifndef __SYCL_DEVICE_ONLY__
+      assert(false && "Type index is invalid");
+#endif
+      return nullptr;
+    }
+
+    return constructHelper<RetT, T...>(TypeIndex, 0);
```

Comment:
I think something like https://github.com/intel/llvm/blob/8eeb60d2ad8f961266c3beec0c68304421928876/sycl/include/sycl/vector.hpp#L1204-L1209 can be used to eliminate the helper.

Reply:
`TypeIndex` is a runtime value here, the function is *not* `constexpr`.
Otherwise, I think I could have used [`tuple_element`](https://en.cppreference.com/w/cpp/utility/tuple/tuple_element) as built-in helper, essentially